pkg/analysis/optionalfields/util.go

```diff
@@ -0,0 +1,431 @@
+	markers := markersAccess.FieldMarkers(field)
+	structMarkers := markersAccess.StructMarkers(structType)
+
+	markers.Insert(structMarkers.UnsortedList()...)
+
+	return markers
+}
+
+// number is a type constraint for numeric types.
+// It allows us to create a generic extraction function for numeric values from markers.
+type number interface {
+	int | float64
+}
+
+// isIntegerZeroValueValid checks if an integer field can have a zero value.
+//
+//nolint:cyclop
+func isNumericZeroValueValid[N number](pass *analysis.Pass, field *ast.Field, markersAccess markershelper.Markers) (bool, bool) {
+	fieldMarkers := markersAccess.FieldMarkers(field)
+
+	minimum, err := getMarkerNumericValueByName[N](fieldMarkers, markers.KubebuilderMinimumMarker)
+	if err != nil && !errors.Is(err, errMarkerMissingValue) {
+		pass.Reportf(field.Pos(), "field %s has an invalid minimum marker: %v", utils.FieldName(field), err)
+		return false, false
+	}
+
+	maximum, err := getMarkerNumericValueByName[N](fieldMarkers, markers.KubebuilderMaximumMarker)
+	if err != nil && !errors.Is(err, errMarkerMissingValue) {
+		pass.Reportf(field.Pos(), "field %s has an invalid maximum marker: %v", utils.FieldName(field), err)
+		return false, false
+	}
+
+	hasGreaterThanZeroMinimum := minimum != nil && *minimum >= 0
+	hasLessThanZeroMaximum := maximum != nil && *maximum <= 0
+	hasCompleteRange := minimum != nil && maximum != nil && *minimum <= *maximum
+
+	return ptr.Deref(minimum, -1) <= 0 && ptr.Deref(maximum, 1) >= 0, hasCompleteRange || hasGreaterThanZeroMinimum || hasLessThanZeroMaximum
+}
+
+// getMarkerNumericValueByName extracts the numeric value from the first instance of the marker with the given name.
+// Works for markers like MaxLength, MinLength, etc.
+func getMarkerNumericValueByName[N number](marker markershelper.MarkerSet, markerName string) (*N, error) {
+	markerList := marker.Get(markerName)
+	if len(markerList) == 0 {
+		return nil, errMarkerMissingValue
+	}
+
+	markerValue, err := getMarkerNumericValue[N](markerList[0])
+	if err != nil {
+		return nil, fmt.Errorf("error getting marker value: %w", err)
+	}
+
+	return &markerValue, nil
+}
+
+// getMarkerNumericValue extracts a numeric value from the default value of a marker.
+// Works for markers like MaxLength, MinLength, etc.
+func getMarkerNumericValue[N number](marker markershelper.Marker) (N, error) {
+	rawValue, ok := marker.Expressions[""]
+	if !ok {
+		return N(0), errMarkerMissingValue
+	}
+
+	value, err := strconv.ParseFloat(rawValue, 64)
+	if err != nil {
+		return N(0), fmt.Errorf("error converting value to number: %w", err)
+	}
+
+	// Convert the float64 value to the desired type N.
+	var numberValue N
+	switch any(&numberValue).(type) {
+	case *int:
+		numberValue = N(int(value))
+	case *float64:
+		numberValue = N(value)
+	default:
+		// Panic here since this error is a programming error.
+		panic(fmt.Sprintf("unsupported number type: %T", numberValue))
+	}
```

Comment:
Is there a particular reason that doing a `return N(value)` is insufficient here?

Having a switch on types is usually an anti-pattern when using generics.

I played around with this and it seems like a `return N(value)` _should_ work - https://go.dev/play/p/U_03gG-Fcl8 (might be a bit naive, but I think this is representative of the scenario here)